VisualPinball.Unity/VisualPinball.Unity/VPT/Flipper/FlipperCorrection.cs

```diff
@@ -0,0 +1,151 @@
+			ref var polarities = ref flipperCorrectionBlob.Polarities;
+			var angleCur = flipperMovementData.Angle;
+			var flipperStrength = flipperStaticData.Strength;
+
+			var ballPosition = ballData.Position;
+			var ballVelocity = ballData.Velocity;
+			var uncorrectedVel = ballVelocity;
+			if (ballVelocity.y > -8F) // ball going down
+			{
+				return;
+			}
+			
+
+			var angleAtFire = flipperMovementData.AngleAtRotateToEndTime;
+			var angleStart = flipperStaticData.AngleStart;
+			var angleEnd = flipperStaticData.AngleEnd;
+
+			var flipPos = flipperStaticData.Position;
+
+			var flipEnd = flipperStaticData.Position;
+			flipEnd.x += math.sin(angleCur) * flipperStaticData.FlipperRadius;
+			flipEnd.y += -math.cos(angleCur) * flipperStaticData.FlipperRadius;
+
+			// Compute ball distance on Flipper (normalized from start to end)
+			//var dir = flipEnd - flipPos;
+			var ballPos = (ballPosition.x - flipPos.x) / (flipEnd.x - flipPos.x);
+
+			// Safety coeffcient: has been disabled in all curves of nFozzy. Not using;
+			float Ycoef = 1F;
+			
+
+			// Normalized flipper course since fire (can be > 1 if rebounding...)
+			float partialFlipCoef = ((angleStart - angleAtFire) / (angleStart - angleEnd));
+			partialFlipCoef = math.abs(partialFlipCoef - 1F);
+
+			// Velocity correction
+			var velCoef = LinearEnvelopeEven(ballPos, ref velocities);
+			var velCoefInit = velCoef;
+			if (partialFlipCoef < 1)
+			{
+				velCoef = PSlope(partialFlipCoef, 0, 1, 1, velCoef);
+			}
+			ballVelocity *= velCoef;
+
+			// Polarity Correction
+			bool isLeft = angleEnd < angleStart; // TODO: better if not classic flippers (trigonometry problems)
+
+			float AddX = LinearEnvelopeEven(ballPos, ref polarities, 0F);
+			if(!isLeft) {
+				AddX = -AddX;
+			}
+			ballVelocity.x += (AddX * Ycoef * partialFlipCoef);
+
+
+
+			// Apply all corrections
+			ballData.Velocity = ballVelocity;
+		}
+
+		// awefull linear interpolations : TODO: replace by AnimationCurve equivalent...
+		private static float PSlope(float x, float x1, float y1, float x2, float y2)  //Set up line via two points, no clamping. Input X, output Y
+		{
+			float m = (y2 - y1) / (y2 - x1);
+			float b = y2 - m * y2;
+			return m * x + b;
+		}
+
+		private static float LinearEnvelope(float xInput, ref BlobArray<float2> curve, float defaultValue = 1F)
+		{
+			if (curve.Length <= 0)
+				return defaultValue;
+
+			if (xInput <= curve[0].x) //Clamp lower
+				return curve[0].y;
+			if (xInput >= curve[curve.Length - 1].x) //Clamp upper
+				return curve[curve.Length - 1].y;
+
+			int L = -1;
+			for (int ii = 1; ii < curve.Length; ii++)    //find active line
+				if (xInput <= curve[ii].x)
+				{
+					L = ii;
+					break;
+				}
+
+			if (L < 0)
+				return defaultValue;
+
+			if (xInput > curve[curve.Length - 1].x) // catch line overrun
+				L = curve.Length - 1;
+
+			float y = PSlope(xInput, curve[L - 1].x, curve[L - 1].y, curve[L].x, curve[L - 1].y);
+
+			return y;
+		}

```

Comment:
This one doesn't seem to be used? Ditch in favor of the next method?

Reply:
This one was the direct port of nFozzy's code. Th eother one is an optimization only valid if the segments of the curve are evenly distributed on x axis. So it will depend on what we plan for the curves to come. It is indeed not used right now as the Blobs are filled evenly splitting the aniamtioncurves.

Reply:
Ok, keeping it for now then.